### PR TITLE
[KYUUBI #502][SPARK-35002][INFRA] Fix the java.net.BindException when testing with Github Action

### PR DESCRIPTION
**Pull request by turboFei**
turboFei [Closes #503](https://github.com/yaooqinn/kyuubi/pull/503)    Test Plan [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=yaooqinn&utm_campaign=badge_info)

### _Why are the changes needed?_
Refer https://github.com/apache/spark/pull/32096, https://github.com/apache/spark/pull/32102,  this PR tries to fix the java.net.BindException when testing with Github Action.

```
SparkOperationSuite:
*** RUN ABORTED ***
  java.net.BindException: Cannot assign requested address: Service 'sparkDriver' failed after 16 retries (on a random free port)! Consider explicitly setting the appropriate binding address for the service 'sparkDriver' (for example spark.driver.bindAddress for SparkDriver) to the correct binding address.
  at sun.nio.ch.Net.bind0(Native Method)
  at sun.nio.ch.Net.bind(Net.java:461)
  at sun.nio.ch.Net.bind(Net.java:453)
  at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:222)
  at io.netty.channel.socket.nio.NioServerSocketChannel.doBind(NioServerSocketChannel.java:134)
  at io.netty.channel.AbstractChannel$AbstractUnsafe.bind(AbstractChannel.java:550)
  at io.netty.channel.DefaultChannelPipeline$HeadContext.bind(DefaultChannelPipeline.java:1334)
  at io.netty.channel.AbstractChannelHandlerContext.invokeBind(AbstractChannelHandlerContext.java:506)
  at io.netty.channel.AbstractChannelHandlerContext.bind(AbstractChannelHandlerContext.java:491)
  at io.netty.channel.DefaultChannelPipeline.bind(DefaultChannelPipeline.java:973)
```

Also transfer FRONTEND_BIND_HOST by connection string to fix similar issue.

```
Cause: java.lang.RuntimeException: org.apache.kyuubi.KyuubiSQLException:org.apache.kyuubi.KyuubiException: Failed to initialize frontend service on fv-az207-19/10.1.1.0:0.
	at org.apache.kyuubi.service.FrontendService.initialize(FrontendService.scala:102)
	at org.apache.kyuubi.service.CompositeService.$anonfun$initialize$1(CompositeService.scala:40)
	at org.apache.kyuubi.service.CompositeService.$anonfun$initialize$1$adapted(CompositeService.scala:40)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at org.apache.kyuubi.service.CompositeService.initialize(CompositeService.scala:40)
	at org.apache.kyuubi.service.Serverable.initialize(Serverable.scala:44)
	at org.apache.kyuubi.engine.spark.SparkSQLEngine.initialize(SparkSQLEngine.scala:49)
	at org.apache.kyuubi.engine.spark.SparkSQLEngine$.startEngine(SparkSQLEngine.scala:105)
	at org.apache.kyuubi.engine.spark.SparkSQLEngine$.main(SparkSQLEngine.scala:118)
	at org.apache.kyuubi.engine.spark.SparkSQLEngine.main(SparkSQLEngine.scala)
```

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/latest/tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Review by yaooqinn — APPROVED** [review]

**Comment by turboFei** [comment]
```
  Cause: java.lang.RuntimeException: org.apache.kyuubi.KyuubiSQLException:org.apache.kyuubi.KyuubiException: Failed to initialize frontend service on fv-az207-19/10.1.1.0:0.
	at org.apache.kyuubi.service.FrontendService.initialize(FrontendService.scala:102)
	at org.apache.kyuubi.service.CompositeService.$anonfun$initialize$1(CompositeService.scala:40)
	at org.apache.kyuubi.service.CompositeService.$anonfun$initialize$1$adapted(CompositeService.scala:40)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at org.apache.kyuubi.service.CompositeService.initialize(CompositeService.scala:40)
	at org.apache.kyuubi.service.Serverable.initialize(Serverable.scala:44)
	at org.apache.kyuubi.engine.spark.SparkSQLEngine.initialize(SparkSQLEngine.scala:49)
	at org.apache.kyuubi.engine.spark.SparkSQLEngine$.startEngine(SparkSQLEngine.scala:105)
	at org.apache.kyuubi.engine.spark.SparkSQLEngine$.main(SparkSQLEngine.scala:118)
	at org.apache.kyuubi.engine.spark.SparkSQLEngine.main(SparkSQLEngine.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.JavaMainApplication.start(SparkApplication.scala:52)
	at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:951)
	at org.apache.spark.deploy.SparkSubmit$$anon$1.run(SparkSubmit.scala:165)
	at org.apache.spark.deploy.SparkSubmit$$anon$1.run(SparkSubmit.scala:163)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1746)
	at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:163)
	at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:203)
	at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:90)
	at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1030)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1039)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
Caused by: java.net.BindException: Cannot assign requested address (Bind failed)
	at java.net.PlainSocketImpl.socketBind(Native Method)
	at java.net.AbstractPlainSocketImpl.bind(AbstractPlainSocketImpl.java:387)
	at java.net.ServerSocket.bind(ServerSocket.java:392)
	at java.net.ServerSocket.<init>(ServerSocket.java:254)
	at org.apache.kyuubi.service.FrontendService.initialize(FrontendService.scala:76)
 See more: /home/runner/work/kyuubi/kyuubi/kyuubi-main/target/work/runner/kyuubi-spark-sql-engine.log.11
  at org.apache.kyuubi.KyuubiSQLException$.apply(KyuubiSQLException.scala:58)
  at org.apache.kyuubi.engine.ProcBuilder.$anonfun$start$1(ProcBuilder.scala:125)
  at java.lang.Thread.run(Thread.java:748)
```

**Comment by turboFei** [comment]
Also set FRONTEND_BIND_PORT env to 127.0.0.1 and have a try.

File: .github/workflows/master.yml [hunk 1]
```diff
@@ -37,6 +37,9 @@ jobs:
           - '-Pspark-3.0 -Dspark.archive.mirror=https://archive.apache.org/dist/spark/spark-3.1.1 -Dspark.archive.name=spark-3.1.1-bin-hadoop2.7.tgz -Dmaven.plugin.scalatest.exclude.tags=org.apache.kyuubi.tags.DataLakeTest'
           - '-Pspark-3.1'
           - '-Pspark-3.1 -Dhadoop.binary.version=3.2'
+    env:
+      SPARK_LOCAL_IP: 127.0.0.1
+      KYUUBI_FRONTEND_BIND_HOST: 127.0.0.1
```
Comment (yaooqinn) [inline .github/workflows/master.yml:42]: How about just add envs and properties in `scalatest-maven-plugin`, add a new env is overkill for me to only fix a CI issue

Comment (turboFei) [reply .github/workflows/master.yml:42]: sounds good

**Comment by turboFei** [comment]
https://github.com/yaooqinn/kyuubi/pull/503/checks?check_run_id=2306833395#step:9:24197

It seems that we have to use System env solution to fix it.

@yaooqinn 
Because the suite is triggered by kyuubi server side, and the property value could not be transferred to engine side.

**Comment by turboFei** [comment]
or transfer it in connection string

File: kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithKyuubiServer.scala [hunk 2]
```diff
@@ -72,5 +72,6 @@ trait WithKyuubiServer extends KyuubiFunSuite {
     super.afterAll()
   }
 
-  protected def getJdbcUrl: String = s"jdbc:hive2://${server.connectionUrl}/;"
+  protected def getJdbcUrl: String =
+    s"jdbc:hive2://${server.connectionUrl}/;${KyuubiConf.FRONTEND_BIND_HOST.key}=127.0.0.1"
```
Comment (turboFei) [inline kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithKyuubiServer.scala:76]: I am not sure whether it could work

Comment (yaooqinn) [reply kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithKyuubiServer.scala:76]: this will work... but OMG... this is a potentical bug..

Comment (turboFei) [reply kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithKyuubiServer.scala:76]: yes, it should be forbid in engine side

**Comment by pan3793** [comment]
`build/dependency.sh` doesn't seem to work as expected, after this get in, I will try to fix it.

File: kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithKyuubiServer.scala (hunk 2 above)
Comment (yaooqinn) [reply kyuubi-main/src/test/scala/org/apache/kyuubi/operation/WithKyuubiServer.scala:76]: Let create an Issue ticket for it, and mark it todo here

File: .github/workflows/master.yml [hunk 3]
```diff
@@ -37,7 +37,8 @@ jobs:
           - '-Pspark-3.0 -Dspark.archive.mirror=https://archive.apache.org/dist/spark/spark-3.1.1 -Dspark.archive.name=spark-3.1.1-bin-hadoop2.7.tgz -Dmaven.plugin.scalatest.exclude.tags=org.apache.kyuubi.tags.DataLakeTest'
           - '-Pspark-3.1'
           - '-Pspark-3.1 -Dhadoop.binary.version=3.2'
-          - '-Pspark-3.2-snapshot -pl :kyuubi-spark-sql-engine,:kyuubi-common,:kyuubi-ha,:kyuubi-zookeeper -Dmaven.plugin.scalatest.exclude.tags=org.apache.kyuubi.tags.DataLakeTest'
```
Comment (pan3793) [inline .github/workflows/master.yml:40]: keep this line

**Review by pan3793 — APPROVED** [review]

**Comment by pan3793** [comment]
Spark switched `127.0.0.1` to `localhost`, see https://github.com/apache/spark/pull/32102

**Comment by pan3793** [comment]
Thanks! Merged into master for v1.2.0.

**Comment by pan3793** [comment]
The issue is reported at https://github.com/actions/virtual-environments/issues/3185
